toolkit/tools/pkg/imagecustomizerlib/customizeboot.go

```diff
@@ -170,40 +159,78 @@ func appendKernelCommandLineArguments(inputGrubCfgContent string, extraCommandLi
 }
 
-func replaceKernelCommandLineArgumentValue(inputGrubCfgContent string, name string, value string) (outputGrubCfgContent string, oldValue string, err error) {
+func findKernelCommandLineArgumentArg(grub2Config string, name string) (int, int, string, error) {
+	const (
+		argGroupIndex      = 2
+		argGroupStartIndex = argGroupIndex*2 + 0
+		argGroupEndIndex   = argGroupIndex*2 + 1
+
+		argValueGroupIndex      = 4
+		argValueGroupStartIndex = argValueGroupIndex*2 + 0
+		argValueGroupEndIndex   = argValueGroupIndex*2 + 1
+	)
 
-	// just in case the input has some special characters
 	escapedName := regexp.QuoteMeta(name)
 
-	// construct the regular expression with one sub-group capturing the old
-	// value
-	argumentRegexString := "(?m)[\\t ]+linux[\\t ]+.*" + escapedName + "=(.*?)[\\t ]"
+	// "linux ... <name>=<value> ..."
+	argumentRegexString := "(?m)[\\t ]+linux[\\t ](.*[\\t ])?(" + escapedName + "(=(\\S*)?)?)[\\t ]"
 	argumentRegex, err := regexp.Compile(argumentRegexString)
 	if err != nil {
-		return "", "", fmt.Errorf("failed to compile regular expression (%s)", argumentRegexString)
+		return 0, 0, "", fmt.Errorf("failed to compile regular expression to find (%s) kernel arg:\n%w", name, err)
 	}
 
-	argumentValueStart := 2
-	argumentValueEnd := 3
-
-	// try to find it
-	match := argumentRegex.FindStringSubmatchIndex(inputGrubCfgContent)
+	match := argumentRegex.FindStringSubmatchIndex(grub2Config)
 	if match == nil {
-		return "", "", fmt.Errorf("failed to find (%s) command line argument in grub2 config file", name)
+		return -1, -1, "", nil
 	}
-	if len(match) <= argumentValueEnd {
-		return "", "", fmt.Errorf("failed to find value for (%s).", name)
+
+	argValue := ""
+	if len(match) > argValueGroupEndIndex {
+		argValueStartIndex := match[argValueGroupStartIndex]
+		argValueEndIndex := match[argValueGroupEndIndex]
+		argValue = grub2Config[argValueStartIndex:argValueEndIndex]
+	}
+
+	argStartIndex := match[argGroupStartIndex]
+	argEndIndex := match[argGroupEndIndex]
+
+	return argStartIndex, argEndIndex, argValue, nil
+}
+
+func replaceKernelCommandLineArgumentValue(inputGrubCfgContent string, name string, value string,
+) (outputGrubCfgContent string, oldValue string, err error) {
+	// Search for the arg.
+	start, end, oldValue, err := findKernelCommandLineArgumentArg(inputGrubCfgContent, name)
+	if err != nil {
+		return "", "", err
 	}
 
-	start := match[argumentValueStart]
-	end := match[argumentValueEnd]
+	if start == -1 {
+		return "", "", fmt.Errorf("failed to find kernel arg value for (%s)", name)
+	}
 
-	oldValue = inputGrubCfgContent[start:end]
-	outputGrubCfgContent = inputGrubCfgContent[:start] + value + inputGrubCfgContent[end:]
+	outputGrubCfgContent = fmt.Sprintf("%s%s=%s%s", inputGrubCfgContent[:start], name, value, inputGrubCfgContent[end:])
 
 	return outputGrubCfgContent, oldValue, nil
 }
 
+func tryRemoveKernelCommandLineArgumentValue(inputGrubCfgContent string, name string) (string, int, error) {
```

Comment:
why is this tryRemove..()? what expectations should caller of this have?

Reply:
Add comment.